### ISSUE TITLE
Unexpected ERROR on undefined variable

### ISSUE THREAD

**Issue by mfalt** [tonyhffong/Lint.jl#129]
When testing my package I get an unexpected error from Lint. I think this is because one on the variables in the expression is not defined in the current file. A minimal example is
```f() = g(myfunc)```
which will throw
```julia
ErrorException("syntax: incomplete: premature end of input")
import Base.+
ErrorException("syntax: incomplete: premature end of input")
import Base.-
ErrorException("syntax: incomplete: premature end of input")
import Base.*
ErrorException("syntax: incomplete: premature end of input")
import Base./
ParseError("invalid operator \"../\"")
import Base../
ErrorException("syntax: incomplete: premature end of input")
import Base.==
ParseError("invalid operator \"..+\"")
import Base..+
ParseError("invalid operator \"..-\"")
import Base..-
ParseError("invalid operator \"..*\"")
import Base..*
ErrorException("syntax: incomplete: premature end of input")
import Base.!=
ERROR: BoundsError: attempt to access 2-element Array{Any,1}:
 :(==)             
 :(promote(a,b)...)
  at index [3]
 in lintfunctioncall at /home/mattias/.julia/v0.4/Lint/src/functions.jl:523
 in lintexpr at /home/mattias/.julia/v0.4/Lint/src/Lint.jl:241
 in lintblock at /home/mattias/.julia/v0.4/Lint/src/blocks.jl:69
 in lintexpr at /home/mattias/.julia/v0.4/Lint/src/Lint.jl:175
 in lintfunction at /home/mattias/.julia/v0.4/Lint/src/functions.jl:275
 in lintexpr at /home/mattias/.julia/v0.4/Lint/src/Lint.jl:183
 in lintstr at /home/mattias/.julia/v0.4/Lint/src/Lint.jl:143
 in lintstr at /home/mattias/.julia/v0.4/Lint/src/Lint.jl:116
 in lintinclude at /home/mattias/.julia/v0.4/Lint/src/Lint.jl:316
 in lintfunctioncall at /home/mattias/.julia/v0.4/Lint/src/functions.jl:397
 in lintexpr at /home/mattias/.julia/v0.4/Lint/src/Lint.jl:241
 in lintblock at /home/mattias/.julia/v0.4/Lint/src/blocks.jl:69
 in lintexpr at /home/mattias/.julia/v0.4/Lint/src/Lint.jl:175
 in lintmodule at /home/mattias/.julia/v0.4/Lint/src/modules.jl:11
 in lintexpr at /home/mattias/.julia/v0.4/Lint/src/Lint.jl:195
 in lintstr at /home/mattias/.julia/v0.4/Lint/src/Lint.jl:143
 in lintstr at /home/mattias/.julia/v0.4/Lint/src/Lint.jl:116
 in lintinclude at /home/mattias/.julia/v0.4/Lint/src/Lint.jl:316
 in lintpkgforfile at /home/mattias/.julia/v0.4/Lint/src/Lint.jl:68
 in lintfile at /home/mattias/.julia/v0.4/Lint/src/Lint.jl:102
 in lintfile at /home/mattias/.julia/v0.4/Lint/src/Lint.jl:91
```
when run with `lintfile("myfile.jl")` but gives the expected result when run as 
```julia
julia> lintstr("f() = g(myfunc)")
1-element Array{Lint.LintMessage,1}:
 none:1 E321 myfunc: use of undeclared symbol
```
I get the same behavior on master. Running Julia 0.4.5.

**Comment by tonyhffong** [comment]
what happens if you include an ```include```
```julia
include("file.jl") # with the definition of myfunc in it
f()=g(myfunc)
```

**Comment by mfalt** [comment]
No it doesn't seem like that helps.
I can also reproduce the same error using for example this code:
```julia
funcname = :myfunc
@eval begin
      function $funcname()
        1
      end
    end

@doc "asd" -> myfunc
```
It is the last line that poses a problem for lint (but not the parser). Lint will only crash if I haven't included the file already. The error is again
```julia
lintfile("testfile.jl")
ErrorException("syntax: incomplete: premature end of input")
import Base.+
ErrorException("syntax: incomplete: premature end of input")
import Base.-
ErrorException("syntax: incomplete: premature end of input")
import Base.*
ErrorException("syntax: incomplete: premature end of input")
import Base./
ParseError("invalid operator \"../\"")
import Base../
ErrorException("syntax: incomplete: premature end of input")
import Base.==
ParseError("invalid operator \"..+\"")
import Base..+
ParseError("invalid operator \"..-\"")
import Base..-
ParseError("invalid operator \"..*\"")
import Base..*
ErrorException("syntax: incomplete: premature end of input")
import Base.!=
ERROR: BoundsError: attempt to access 2-element Array{Any,1}:
 :(==)             
 :(promote(a,b)...)
  at index [3]
 in lintfunctioncall at /home/mattias/.julia/v0.4/Lint/src/functions.jl:523
 in lintexpr at /home/mattias/.julia/v0.4/Lint/src/Lint.jl:241
 in lintblock at /home/mattias/.julia/v0.4/Lint/src/blocks.jl:69
 in lintexpr at /home/mattias/.julia/v0.4/Lint/src/Lint.jl:175
 in lintfunction at /home/mattias/.julia/v0.4/Lint/src/functions.jl:275
 in lintexpr at /home/mattias/.julia/v0.4/Lint/src/Lint.jl:183
 in lintstr at /home/mattias/.julia/v0.4/Lint/src/Lint.jl:143
 in lintstr at /home/mattias/.julia/v0.4/Lint/src/Lint.jl:116
 in lintinclude at /home/mattias/.julia/v0.4/Lint/src/Lint.jl:316
 in lintfunctioncall at /home/mattias/.julia/v0.4/Lint/src/functions.jl:397
 in lintexpr at /home/mattias/.julia/v0.4/Lint/src/Lint.jl:241
 in lintblock at /home/mattias/.julia/v0.4/Lint/src/blocks.jl:69
 in lintexpr at /home/mattias/.julia/v0.4/Lint/src/Lint.jl:175
 in lintmodule at /home/mattias/.julia/v0.4/Lint/src/modules.jl:11
 in lintexpr at /home/mattias/.julia/v0.4/Lint/src/Lint.jl:195
 in lintstr at /home/mattias/.julia/v0.4/Lint/src/Lint.jl:143
 in lintstr at /home/mattias/.julia/v0.4/Lint/src/Lint.jl:116
 in lintinclude at /home/mattias/.julia/v0.4/Lint/src/Lint.jl:316
 in lintpkgforfile at /home/mattias/.julia/v0.4/Lint/src/Lint.jl:68
 in lintfile at /home/mattias/.julia/v0.4/Lint/src/Lint.jl:102
 in lintfile at /home/mattias/.julia/v0.4/Lint/src/Lint.jl:91

```

**Comment by tonyhffong** [comment]
You didn't mention you hid the definition of `myfunc` inside a macro. No, Lint doesn't evaluate inside `@eval` and therefore it would not have known that myfunc has been conjured into existence.

**Comment by mfalt** [comment]
This problem occurs even if I define the function in the normal way in another file (which was my original problem). It would be nice if Lint was able to output something like the error you get when running `lintstr`
`none:1 E321 myfunc: use of undeclared symbol`
instead of crashing, since this makes it unusable when either running `lintfile` or `lintpkg`.

**Comment by mfalt** [comment]
Do you agree that this is a bug?
If you don't have time to fix it then I might be able to take a stab at it, if you could direct me where to look.

**Comment by Michael-Klassen** [comment]
@mfalt I agree this is a bug. Anytime the linter crashes is a bug. The correct output would be `E321` as you stated above.

If you want to take a stab at it I would recommend you start where the `BoundsError` occurs. We are probably missing a bounds check. If Lint can't find a symbol in your code or your included files it will lint base to try and find the symbol. It is likely the code that is causing the crash is some new code added to base that Lint is having troubles with. It would be useful to track down the actual code causing the error.

**Comment by Michael-Klassen** [comment]
I was not able to reproduce the exception. I'm also running Julia Version 0.4.5.

**Comment by mfalt** [comment]
Interesting, I think I was able to reproduce it on 2 different linux machines with 0.4.2. I will try again.

**Comment by mfalt** [comment]
The above result is for both master and the latest tagged version.

**Comment by Michael-Klassen** [comment]
I'm using a mac. Maybe it's linux specific. I'll fire up a VM.

**Comment by Michael-Klassen** [comment]
Works for me in my Debian VM.

**Comment by Michael-Klassen** [comment]
Can you add a print statement on line 522 of  `functions.jl` and see what is in `ex`?

**Comment by mfalt** [comment]
From the following code:
```julia
        if ex.args[1] in COMPARISON_OPS
            #reuse lintcomparison by synthetically construct the expr
            println("Pre")
            println(ex)
            println("Post")
            lintcomparison( Expr( :comparison, ex.args[2], ex.args[1], ex.args[3] ), ctx )
        end
```
I get the output
```julia
...
...
Pre
(==)(promote(a,b)...)
Post
ERROR: BoundsError: attempt to access 2-element Array{Any,1}:
 :(==)             
 :(promote(a,b)...)
  at index [3]
 in lintfunctioncall at /home/mattias/.julia/v0.4/Lint/src/functions.jl:526
 in lintexpr at /home/mattias/.julia/v0.4/Lint/src/Lint.jl:241
 in lintblock at /home/mattias/.julia/v0.4/Lint/src/blocks.jl:69
 in lintexpr at /home/mattias/.julia/v0.4/Lint/src/Lint.jl:175
 in lintfunction at /home/mattias/.julia/v0.4/Lint/src/functions.jl:275
 in lintexpr at /home/mattias/.julia/v0.4/Lint/src/Lint.jl:183
 in lintstr at /home/mattias/.julia/v0.4/Lint/src/Lint.jl:143

```

**Comment by Michael-Klassen** [comment]
Ok, now can you also print `ctx.file`  `ctx.line`  `ctx.lineabs`  `ctx.scope`  `ctx.path` .

**Comment by mfalt** [comment]
So this is extremely weird. Looking at the output I realized that the error depends on which folder I'm putting the file in. I was just putting the file in the ControlSystems folder (in which I'm writing code and discovered the error the first time) and the error seems to occur in a completely different file in that package [transferfunction.jl](https://github.com/mfalt/ControlSystems.jl/blob/master/src/types/transferfunction.jl) which is neither included, loaded or refereed to in any way in the code
```julia
f() = g(myfunc)
```
in the file `test.jl`.
```julia
julia> lintfile("test.jl")
ErrorException("syntax: incomplete: premature end of input")
import Base.+
ErrorException("syntax: incomplete: premature end of input")
import Base.-
ErrorException("syntax: incomplete: premature end of input")
import Base.*
ErrorException("syntax: incomplete: premature end of input")
import Base./
ParseError("invalid operator \"../\"")
import Base../
ErrorException("syntax: incomplete: premature end of input")
import Base.==
ParseError("invalid operator \"..+\"")
import Base..+
ParseError("invalid operator \"..-\"")
import Base..-
ParseError("invalid operator \"..*\"")
import Base..*
ErrorException("syntax: incomplete: premature end of input")
import Base.!=
ctx.file
UTF8String "/home/mattias/.julia/v0.4/ControlSystems/src/types/transferfunction.jl"
ctx.line
Int64 0
ctx.lineabs
Int64 116
ctx.scope
UTF8String "=="
ctx.path
UTF8String "/home/mattias/.julia/v0.4/ControlSystems/src/types"
ex
Expr 
  head: Symbol call
  args: Array(Any,(2,))
    1: Symbol ==
    2: Expr 
      head: Symbol ...
      args: Array(Any,(1,))
        1: Expr 
          head: Symbol call
          args: Array(Any,(3,))
          typ: Any
      typ: Any
  typ: Any
Post
ERROR: BoundsError: attempt to access 2-element Array{Any,1}:
 :(==)             
 :(promote(a,b)...)
  at index [3]
 in lintfunctioncall at /home/mattias/.julia/v0.4/Lint/src/functions.jl:536
 in lintexpr at /home/mattias/.julia/v0.4/Lint/src/Lint.jl:241
 in lintblock at /home/mattias/.julia/v0.4/Lint/src/blocks.jl:69
 in lintexpr at /home/mattias/.julia/v0.4/Lint/src/Lint.jl:175
 in lintfunction at /home/mattias/.julia/v0.4/Lint/src/functions.jl:275
 in lintexpr at /home/mattias/.julia/v0.4/Lint/src/Lint.jl:183
 in lintstr at /home/mattias/.julia/v0.4/Lint/src/Lint.jl:143

```

**Comment by Michael-Klassen** [comment]
Thanks. I'll submit a pull request with a fix on Monday.
